### PR TITLE
[euphoria-core] #259 Hints are not runtime specific.

### PR DESCRIPTION
**Pull request by mareksimunek**
#259  first part of unifying hints

implemented 
```java
public Dataset<OUT> output(Hint... hints)
```
Its breaking change because `withHints` on Join operator no longer exists

### CODE REVIEW

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/Dataset.java [hunk 1]
```diff
@@ -20,17 +20,20 @@
...
-import javax.annotation.Nullable;
 
 /**
  * A dataset abstraction.
  *
  * @param <T> type of elements of this data set
  */
 @Audience(Audience.Type.CLIENT)
-public interface Dataset<T> extends Serializable {
+public interface Dataset<T> extends OutputHintAware<OutputHint>, Serializable {
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/Dataset.java]: I don't understand purpose of this interface. It looks odd in the context of client API. We should definitely keep this set on operator, not dataset, because operator is not (directly) part of client API.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/Datasets.java [hunk 2]
```diff
@@ -39,9 +42,9 @@
    * @return a dataset representing the output of the given operator
    */
   public static <IN, OUT> Dataset<OUT> createOutputFor(
-      Flow flow, Dataset<IN> input, Operator<IN, OUT> op) {
+      Flow flow, Dataset<IN> input, Operator<IN, OUT> op, Set<OutputHint> outputHints) {
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/Datasets.java]: If we set output hints on operator, there is no need for additional parameter here. On Beam branch, I already further simplified this method, because the parameter `flow` is redundant, because it can be retrieved from both `input` and `op`.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/InputDataset.java [hunk 3]
```diff
@@ -20,8 +20,12 @@
...
-import java.util.Collection;
+import cz.seznam.euphoria.core.client.operator.hint.OutputHint;
+
 import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
 
 /**
  * {@code PCollection} that is input of a {@code Flow}.
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/InputDataset.java]: Although this is not part of this PR, can we fix the comment here?

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/InputDataset.java [hunk 4]
```diff
@@ -57,6 +61,15 @@ public void persist(DataSink<T> sink) {
...
   }
 
+  /**
+   * Input Dataset doesn't have hints
+   * @return empty set
+   */
+  @Override
+  public Set<OutputHint> getHints() {
+    return Collections.emptySet();
+  }
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/InputDataset.java]: Another +1 for moving hints to operator, this useless method will disappear. Moreover, the hints are called `OutputHint`, and `Dataset` has no "output". Operator does.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Builders.java [hunk 5]
```diff
@@ -106,24 +105,13 @@
      *
      * @return the dataset representing the new operator's output
      */
-    default Dataset<V> outputValues() {
+    default Dataset<V> outputValues(OutputHint... outputHints) {
       return MapElements
           .named("extract-values")
           .of(output())
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Builders.java]: The hints should be set here I think.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Filter.java [hunk 6]
```diff
@@ -134,8 +137,9 @@ public static OfBuilder named(String name) {
 
   final UnaryPredicate<IN> predicate;
 
-  Filter(String name, Flow flow, Dataset<IN> input, UnaryPredicate<IN> predicate) {
-    super(name, flow, input);
+  Filter(String name, Flow flow, Dataset<IN> input, UnaryPredicate<IN> predicate, Set<OutputHint>
+      outputHints) {
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Filter.java]: This indentation is very extraordinary. :-D

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/AssignEventTime.java [hunk 7]
```diff
@@ -87,9 +90,10 @@
     }
 
     @Override
-    public Dataset<IN> output() {
+    public Dataset<IN> output(OutputHint... outputHints) {
       Flow flow = input.getFlow();
-      AssignEventTime<IN> op = new AssignEventTime<>(name, flow, input, eventTimeFn);
+      AssignEventTime<IN> op = new AssignEventTime<>(name, flow, input, eventTimeFn, Sets.newHashSet
+          (outputHints));
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/AssignEventTime.java]: Please place the opening parentheses on previous line in this case.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/FlatMap.java [hunk 8]
```diff
@@ -166,9 +170,10 @@
     }
 
     @Override
-    public Dataset<OUT> output() {
+    public Dataset<OUT> output(OutputHint... outputHints) {
       Flow flow = input.getFlow();
-      FlatMap<IN, OUT> map = new FlatMap<>(name, flow, input, functor, evtTimeFn);
+      FlatMap<IN, OUT> map = new FlatMap<>(name, flow, input, functor, evtTimeFn, Sets.newHashSet
+          (outputHints));
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/FlatMap.java]: Opening parentheses placement.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/ReduceByKey.java [hunk 9]
```diff
@@ -426,6 +443,11 @@ public boolean isCombinable() {
     return valueExtractor;
   }
 
+  @Override
+  public Dataset<Pair<KEY, OUT>> output(OutputHint... outputHints) {
+    return output();
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/ReduceByKey.java]: This looks odd. Why two methods, when the varargs version can handle the zero argument version as well?

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/hint/OutputHintAware.java [hunk 10]
```diff
@@ -13,18 +13,19 @@
...
-package cz.seznam.euphoria.core.client.operator;
+package cz.seznam.euphoria.core.client.operator.hint;
 
 import cz.seznam.euphoria.core.annotation.audience.Audience;
+
 import java.util.Set;
 
 @Audience(Audience.Type.INTERNAL)
-public interface HintAware<HINT extends Hint> {
+public interface OutputHintAware<HINT extends OutputHint> {
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/hint/OutputHintAware.java]: I would drop this whole interface.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/hint/SizeHint.java [hunk 11]
```diff
@@ -13,11 +13,15 @@
...
-import cz.seznam.euphoria.core.annotation.audience.Audience;
-
-@Audience(Audience.Type.INTERNAL)
-public interface JoinHint extends Hint {
+package cz.seznam.euphoria.core.client.operator.hint;
 
+/**
+ * Extra information for runner about Dataset size
+ */
+public enum SizeHint implements OutputHint {
```
Comment (je-ik) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/hint/SizeHint.java]: Please use @Audience annotation here.

File: euphoria-core/src/test/java/cz/seznam/euphoria/core/client/operator/JoinTest.java [hunk 12]
```diff
@@ -198,51 +198,58 @@ public void testBuild_Windowing() {
...
 
-    Join join = (Join) flow.operators().iterator().next();
-    assertTrue(join.getHints().contains(new TestHint()));
-    assertTrue(join.getHints().contains(new TestHint2()));
-    assertEquals(2, join.getHints().size());
+    Join join = (Join) flow.operators().stream().filter(op -> op instanceof Join).findFirst().get();
+    assertTrue(join.listInputs().stream().anyMatch(input -> ((Dataset) input).getHints().contains(new
+        TestHint())));
+    assertTrue(join.listInputs().stream().anyMatch(input -> ((Dataset) input).getHints().contains(new
+        TestHint2())));
```
Comment (je-ik) [inline euphoria-core/src/test/java/cz/seznam/euphoria/core/client/operator/JoinTest.java]: These two lines wrap strangely.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Builders.java (hunk 5 above)
Comment (mareksimunek) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Builders.java]: It shouldn't, output Dataset is with values. There would hint get lost. (last operator is MapElements). I added test.

Comment (je-ik) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Builders.java]: Why would the hint get lost? The graph can be arbitrarily complicated and as a general rule, this should yeild the same runtime behavior:
```java
 Dataset input = ...;
 Dataset<Pair> output = ReduceByKey.of(input)
   ...
   .output(hints);
 MapElements.of(input)
   .using(Pair::getSecond)
   .output()
   .persist(...)
```
it should behave the same way as
```java
 Dataset input = ...;
 ReduceByKey.of(input)
    ...
    .outputValues(hints)
    .persist(...)
```
That is because non-shuffle operators should forward hints to downstream Datasets (unless overridden). That is due to the construction of the hints. The operator that the hint applies to is really `RBK`, not the `MapElements`.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/Dataset.java (hunk 1 above)
Comment (mareksimunek) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/Dataset.java]: interface removed. Hints are now on operator

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/Datasets.java (hunk 2 above)
Comment (mareksimunek) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/Datasets.java]: Hints are now on operator

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/InputDataset.java (hunk 3 above)
Comment (mareksimunek) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/client/dataset/InputDataset.java]: fixed in #271

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Builders.java (hunk 5 above)
Comment (mareksimunek) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Builders.java]: After discussion, for now `Hints` doesn't propagate downstream in flow. Because we don't know if `Hint` will be still valid in next operator.

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/FlatMap.java (hunk 8 above)
Comment (mareksimunek) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/FlatMap.java]: fixed

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/hint/OutputHintAware.java (hunk 10 above)
Comment (mareksimunek) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/hint/OutputHintAware.java]: removed

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/hint/SizeHint.java (hunk 11 above)
Comment (mareksimunek) [reply euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/hint/SizeHint.java]: added

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/ReduceByKey.java [hunk 13]
```diff
@@ -395,9 +399,23 @@ public static OfBuilder named(String name) {
     this(
         name, flow, input, keyExtractor, valueExtractor,
         windowing, (ReduceFunctor<VALUE, OUT>) toReduceFunctor(reducer),
-        null);
+        null, Collections.emptySet());
   }
 
+  @SuppressWarnings("unchecked")
+  ReduceByKey(String name,
```
Comment (dmvk) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/ReduceByKey.java]: i would prefer sticking with two constructors

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/ReduceStateByKey.java [hunk 14]
```diff
@@ -355,9 +359,21 @@ public static OfBuilder named(String name) {
...
                    StateFactory<VALUE, OUT, STATE> stateFactory,
-                   StateMerger<VALUE, OUT, STATE> stateMerger)
-  {
-    super(name, flow, input, keyExtractor, windowing);
+                   StateMerger<VALUE, OUT, STATE> stateMerger) {
+    this(name, flow, input, keyExtractor, valueExtractor, windowing, stateFactory, stateMerger,
+        Collections.emptySet());
+  }
+
+  ReduceStateByKey(String name,
```
Comment (dmvk) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/ReduceStateByKey.java]: same here

File: euphoria-core/src/test/java/cz/seznam/euphoria/core/client/operator/JoinTest.java [hunk 15]
```diff
@@ -198,74 +198,58 @@ public void testBuild_Windowing() {
...
+        .outputValues(SizeHint.FITS_IN_MEMORY);
+
+    assertTrue(outputDataset.getProducer().getHints().contains(SizeHint.FITS_IN_MEMORY));
+
+    Join join = (Join) flow.operators().stream().filter(op -> op instanceof Join).findFirst().get();
+    assertTrue(join.listInputs().stream().anyMatch(input ->
+        ((Dataset) input).getProducer().getHints().contains(new Util.TestHint())));
+    assertTrue(join.listInputs().stream().anyMatch(input -> 
+        ((Dataset) input).getProducer().getHints().contains(new Util.TestHint2())));
+    assertEquals(2, ((Dataset) join.listInputs().stream().findFirst().get()).getProducer().getHints().size());
```
Comment (dmvk) [inline euphoria-core/src/test/java/cz/seznam/euphoria/core/client/operator/JoinTest.java]: I reckon multilining this would be beneficial :)

File: euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/SumByKey.java [hunk 16]
```diff
@@ -223,17 +227,26 @@ public static OfBuilder named(String name) {
...
+    super(name, flow, input, keyExtractor, windowing, outputHints);
     this.valueExtractor = valueExtractor;
   }
 
   @Override
   public DAG<Operator<?, ?>> getBasicOps() {
     ReduceByKey<IN, KEY, Long, Long, W> reduceByKey =
         new ReduceByKey<>(getName(), input.getFlow(), input,
-        keyExtractor, valueExtractor, windowing, Sums.ofLongs());
+        keyExtractor, valueExtractor, windowing, Sums.ofLongs(), this.getHints());
```
Comment (dmvk) [inline euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/SumByKey.java]: we should stick with the same coding style :) Can we remove `this.`?
